common/lib/types/protocolmessage.js

```diff
@@ -71,6 +73,30 @@ var ProtocolMessage = (function() {
 		return this.flags && (this.flags & flags.MODE_ALL);
 	};
 
+	ProtocolMessage.prototype.encodeModesToFlags = function(modes) {
+		var self = this;
+		Utils.arrForEach(modes, function(mode) {
+			self.setFlag(mode);
+		});
+	};
```

Comment:
Now we're exposing this as a public api we should do least some kind of error checking on the mode array the user passes in, to prevent typos (e.g. setFlag throws if given a string it doesn't recognise)

Reply:
Does this validation cover all scenarios you have in mind - https://github.com/ably/ably-js/pull/613/files#diff-ee8e6cae22c16e2972f2b341a26d2ed0R77-R96?

Reply:
I'm afraid not, sorry. I made the comment on `ProtocolMessage.prototype.encodeModesToFlag` (and mentioned `setFlag`) for a reason: part of the point of having `options.modes` as a thing (as opposed to just using params.modes) is that you can get clientside validation. The client library has a list of all the modes it knows how to encode in the flags field, and it can reject a mode not in that list. You haven't attempted to do that -- you've added some validation that checks someone hasn't passed as an empty string on a number, but that's kinda missing the point -- we have the ability to statically check that someone hasn't typoed 'subscribe' as 'subscbe', and we should be doing that. (And can be a union type in the type definitions file)

Reply:
Fixed already.